submissions/api.py

```diff
@@ -373,6 +383,40 @@ def get_submissions(student_item_dict, limit=None):
     return SubmissionSerializer(submission_models, many=True).data
 
 
+def get_all_submissions(course_id, item_id, item_type, read_replica=True):
+    """
+    For the given item, get the most recent submission for every student who has submitted.
+
+    This may return a very large result set! It is implemented as a generator for efficiency.
+    It yields namedtuples with the following fields:
+        student_id
+        attempt_number
+        submitted_at
+        created_at
+        answer
+
+    Cannot fail unless there's a database error, but may return an empty list.
+    """
+    query = Submission.objects.filter(
+        student_item__course_id=course_id,
+        student_item__item_id=item_id,
+        student_item__item_type=item_type,
+    ).values_list(
+        'student_item__student_id', 'attempt_number', 'submitted_at', 'created_at', 'raw_answer'
+    ).order_by('student_item__student_id', '-created_at')
+
+    last_student_id = None
+    for row in query.all():
+        # We cannot use SELECT DISTINCT ON because it's PostgreSQL only, so unfortunately
+        # our results may contain every entry of each student, not just the most recent.
+        if row[0] == last_student_id:
+            continue  # Skip this row; it's an old submission for a student we've already included
+        last_student_id = row[0]
+        # Parse 'raw_answer' as JSON:
+        answer = json.loads(row[4])
```

Comment:
According to [a comment in `models.py`](https://github.com/edx/edx-submissions/blob/835b79040/submissions/models.py#L101), `raw_answer` is supposed to be JSON serialized.

I do like the idea of deferring the JSON parsing though.